### DISCUSSION TITLE
romcal v3 country calenders need help.

### DISCUSSION THREAD

**Discussion by pvanthony** [romcal/romcal#416] [Q&A]
Need help with v3 country calenders.

Need to convert my current Singapore calendar to v3 calendar. Looked at the doc but it says it is only for v2.

Here is the current Singapore calendar.
```
import dayjs from 'dayjs';

import * as Locales from '@romcal/lib/locales';
import { LiturgicalColors } from '@romcal/constants/liturgical-colors/liturgical-colors.enum';
import { LiturgicalDayInput } from '@romcal/models/liturgical-day/liturgical-day.types';
import { RomcalConfig, RomcalConfigInCalendarDef } from '@romcal/models/config/config.model';
import { Ranks } from '@romcal/constants/ranks/ranks.enum';
import { Titles } from '@romcal/constants/titles/titles.enum';

const defaultConfig: RomcalConfigInCalendarDef | undefined = undefined;

const dates = async (config: RomcalConfig): Promise<Array<LiturgicalDayInput>> => {
  const year = config.year;
  const _dates: Array<LiturgicalDayInput> = [
    {
      key: 'martha_of_bethany_mary_of_bethany_and_lazarus_of_bethany',
      extend: true,
      prioritized: true,
    },
    {
      key: 'our_lady_of_sorrows',
      extend: true,
      prioritized: true,
    },
    {
      key: 'therese_of_the_child_jesus_and_the_holy_face_of_lisieux_virgin',
      extend: true,
      prioritized: true,
    },
    {
      key: 'guardian_angels',
      extend: true,
      prioritized: true,
    },
    {
      key: 'francis_xavier_priest',
      extend: true,
      prioritized: true,
    },
    {
      key: 'barnabas_apostle',
      extend: true,
      prioritized: true,
    },
    {
      key: 'passion_of_saint_john_the_baptist',
      extend: true,
      prioritized: true,
    },
  ];

  // Get localized liturgical day names
  return await Locales.localizeDates(_dates);
};

export { dates, defaultConfig };

``` 

I looked at the other countries but I do not understand the code. For example what does the following mean?

```
Precedences.ProperSolemnity_PrincipalPatron_4a

precedence: Precedences.ProperMemorial_11b

precedence: Precedences.OptionalMemorial_12
```
Got the above from india.ts

Or where can I go read up about the new syntax?

**Comment by emagnier** [comment]
Hello @pvanthony, your calendar definition should look like this:
```ts
import { Precedences } from '../constants/precedences';
import { CalendarDef } from '../models/calendar-def';
import { Inputs } from '../types/calendar-def';

export class Singapore extends CalendarDef {
  inputs: Inputs = {
    martha_of_bethany_mary_of_bethany_and_lazarus_of_bethany: {
      // ...put anything that needs to be specific to this calendar
    },

    our_lady_of_sorrows: {
      // ...put anything that needs to be specific to this calendar
      // for example, if you want to override the precedence to a solemnity
      precedence: Precedences.ProperSolemnity_PrincipalPatron_4a,
    },

    therese_of_the_child_jesus_and_the_holy_face_of_lisieux_virgin: {
      // ...put anything that needs to be specific to this calendar
    },

    guardian_angels: {
      // ...put anything that needs to be specific to this calendar
    },

    francis_xavier_priest: {
      // ...put anything that needs to be specific to this calendar
    },

    barnabas_apostle: {
      // ...put anything that needs to be specific to this calendar
    },

    passion_of_saint_john_the_baptist: {
      // ...put anything that needs to be specific to this calendar
    },
  };
}
```

However, in your original file, you don't seem to be modifying any items, apart from adding a `prioritized: true`.

Using the `prioritized` property was a hack to bypass some limitations of romcal v1 and v2. Since then, in romcal v3 the management of precedence rules has been completely refactored, and therefore this `prioritized` property, which no longer needs to be, has been removed.

What was your need to use the `prioritized` property? Are you using this calendar for a religious community, or a parish? Because in your case, it's most likely rather the precedence rules that need to be adjusted for your needs.

**Reply by tukusejssirs** [reply]
While @emagnier is right (albeit he forgot about `parentCalendar`, see [this](https://github.com/romcal/romcal/blob/2933b1d8d856eda3117bc79972756fe535f10d1c/lib/particular-calendars/argentina.ts#L9), which should be to `Asia` for `Singapore` national calendar), I think the issue lays in what celebrations should be part of the Singaporean calendar.

First of all, you nees to understand the calender inheritance (see [this comment](https://github.com/romcal/romcal/issues/48#issuecomment-565650048)). Basically, celebrations that are defined in a more general calendar should not be defined in a more particular one, unless the celebration is changed or dropped. If a celebration is changed, only the changed properties should be added to the celebration in the more particular calendar.

If a celebration defined in the GRC or Asian continental calendar is celebrated in the national calendar without any change (i.e. it is celebrated on the same date, with the same titles, with the same rank), it should not be defined in `singapore.ts` at all.

Also don’t forget to add [`particularConfig`](https://github.com/romcal/romcal/blob/dev/lib/particular-calendars/france.ts#L10:L14).

> For example what does the following mean?
> 
> ```ts
> Precedences.ProperSolemnity_PrincipalPatron_4a
> 
> precedence: Precedences.ProperMemorial_11b
> 
> precedence: Precedences.OptionalMemorial_12
> ```

These are the new celebration ranks; look [here](https://github.com/romcal/romcal/blob/dev/lib/constants/precedences.ts) for descriptons and all available options. It is based on the `Table of Liturgical Days` from [_General Norms for the Liturgical Year and the Calendar_ ](https://www.archbalt.org/wp-content/uploads/2017/04/GeneralNormsLiturgicalYear.pdf) (pp7-8 in the PDF).

**Reply by pvanthony** [reply]
An Asia calendar would be great. Could not find Asia calendar in romcal/lib/particular-calendar folder? Is it something that needs to be installed?

**Reply by pvanthony** [reply]
> What was your need to use the `prioritized` property? Are you using this calendar for a religious community, or a parish? Because in your case, it's most likely rather the precedence rules that need to be adjusted for your needs. 

Yes. We do not need to change the calendar. Just need correct one from the array to be shown.

I will study the links and do the precedence settings.

Once again thanks to @tukusejssirs and @emagnier for the advice and example code.

**Reply by tukusejssirs** [reply]
> An Asia calendar would be great. Could not find Asia calendar in romcal/lib/particular-calendar folder? Is it something that needs to be installed?

You’re right, I haven’t check if there is one. We need to create one.

A continental calendar, albeit it is not used separately, is used in order to deduplicate celebrations celebrated everywhere on a continent from the regional/national/diocesan/parish calendars. It contains mostly celebrations of patrons of a particular continent.

According to [Wikipedia](https://en.wikipedia.org/wiki/Patron_saints_of_places#Continents), there is only a single patron of Asia (St Francis Xavier), therefore this celebration should be placed in `asia.ts` calendar. St John the Evangelist is the patron saint of Asia Minor, which should be included in a regional calendar (e.g. `asia.asia-minor.ts`; maybe @emagnier has different idea, but I think we should rename all national calendars to contain the continent name if we want to follow the diocesan calendar name format, although that might be unnecessary, as we could create folders for that with `index.ts` barrel files for each).

Anyway, it might a good idea to check your national calendar for any celebration(s) of continental significance (like patrons of Asia). Each such celebration should be moved to `asia.ts`. Based on the Wikipedia article above, I believe there will be only one (the Memorial of St Francis Xavier), but better be sure.

I’d be very grateful if you could provide me with the General Roman Calendar from the ~~Tamil/~~ English ~~/whatever~~ translation of the Roman Missal used in Singapore (based on [this] I know that you use an English translation of the Roman Missal modified for usage in Malaysia, Singapore and Brunei). Example of the calendar is [here (pp119-130)](https://drive.google.com/file/d/1QZq2h6GeTMsRwwaspRjhHitQymkeizJs/view) ([direct link to the page](https://drive.google.com/uc?export=view&id=1QZq2h6GeTMsRwwaspRjhHitQymkeizJs#page=10), but slow to load). For citation purposes, it’d be awesome to include the front cover, title page and title page verso too. You could simply take some pictures from your parish Missal (OFC, after consulting your priest and/or sacristan). Although I can’t read Tamil, it would serve us for documentation purposes and I can still use Google Translate to get the basic idea of what celebrations are used in your country.

> Yes. We do not need to change the calendar. Just need correct one from the array to be shown.

Are you sure? If [your calendar example from the OP](https://github.com/romcal/romcal/discussions/416#discussion-4823504) is the _actual_ calendar you used in `romcal@2`, it might be wrong, thus needs to be corrected. If that is not the actual (full) calendar you currently use, could you share it (ideally as a file attachment) with us please?

Now, I want to emphasise that no celebration which was inscribed in a more general calendar should be inscribed in a more particular calendar, unless there is a change in the celebration.

Here’s the list of calendar ranks ordered by particularity from more general to more particular (not calendar ranks need to be used for each calendar, but the GRC always is used):
- the GRC;
- continental calendar;
- regional calendar;
- national calendar;
- diocesan calendar;
- parish calendar.

As it was already stated by @emagnier, you should not use `extend` and `prioritized` in `romcal@3`.

Anyway, if you want us help you create `singore.ts` national calendar, we need the Singaporean Calendar Proper (a list of celebrations inscribed in the calendar). If you include all celebrations (even those inscribed in the GRC), I could help you sort out what should be inscribed in which calendar. If you have no access to the Proper, could you share with a liturgical calendar / Ordo (ideally for multiple years; like [this](https://www.amazon.com/Catholic-directory-singapore-malaysia-brunei/dp/B00DODRIGE))? That is one way to (re-) create the Proper.

**Reply by tukusejssirs** [reply]
I have found [this calendar](https://www.catholic.sg/events/catholic-liturgical-calendar/) on the website of the Roman Catholic Archdiocese of Singapore, which _seems_ to be sourced from [Universalis](https://universalis.com/asia.singapore/calendar.htm).

Based on Universalis Singaporean National Calendar for 2023, I have created the following list of celebrations that are inscribed in the Calendar Proper (as I have created this list in hurry, I might miss some additional changes, like changing the celebration rank level). Note, however, that I consider Universals a third-party, unofficial source, therefore getting access to an official (first-party) source like a decree by the Holy See or the GRC from the Roman Missal would be nice to have. Also note that [National calendars of the Roman Rite](https://en.wikipedia.org/wiki/National_calendars_of_the_Roman_Rite) on Wikipedia does not contain a list of celebrations celebrated in Singapore.

| Date        | Name                                               | Rank              | Change type | Notes                                                                                                                                                         |
|:------------|:---------------------------------------------------|:------------------|:------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 19 June     | Saint Philip Minh, Priest, and Companions, Martyrs | memorial          | addition    | Was he one of the [Vietnamese Martyrs](https://en.wikipedia.org/wiki/Vietnamese_Martyrs)? Was `Philip Minh Van Phan` or `Philip Minh Van Doan` his full name? |
| 5 September | Saint Teresa of Calcutta, Religious                | optional memorial | addition    | romcal l10n ID: `teresa_of_calcutta_religious`                                                                                                                |

**Reply by pvanthony** [reply]
Please give me a while. I am waiting for reply from Friar Julian.

**Reply by tukusejssirs** [reply]
Take any time you want! :wink:

**Reply by pvanthony** [reply]
You all were right. There was only one change for now. About the other Saint Philip Minh and Saint Mother Teresa, we are holding off until we get some confirmation from the archdiocese.

Here is the file used for singapore.ts

```
import { Precedences } from '../constants/precedences';
import { CalendarDef } from '../models/calendar-def';
import { Inputs } from '../types/calendar-def';

export class Singapore extends CalendarDef {
  inputs: Inputs = {
    francis_xavier_priest: {
      precedence: Precedences.ProperSolemnity_PrincipalPatron_4a,
      dateDef: { month: 12, date: 3 },
    },
  };
}

```

Thank all for helping with the code for Singapore.

We are still working on the Tamil translation.
